numpy/core/numeric.py

```diff
@@ -1665,15 +1673,36 @@ def indices(dimensions, dtype=int):
     Note that it would be more straightforward in the above example to
     extract the required elements directly with ``x[:2, :3]``.
 
+    If sparse is set to true, the grid will be returned in a sparse
+    representation.
+
+    >>> i, j = np.indices((2, 3), sparse=True)
+    >>> i.shape
+    (2, 1)
+    >>> j.shape
+    (1, 3)
+    >>> i        # row indices
+    array([[0],
+           [1]])
+    >>> j        # column indices
+    array([[0, 1, 2]])
+
     """
     dimensions = tuple(dimensions)
     N = len(dimensions)
     shape = (1,)*N
-    res = empty((N,)+dimensions, dtype=dtype)
+    if sparse:
+        res = tuple()
```

Comment:
Since for the dense case `res` is guaranteed to be much bigger than the sparse arrays that constitute it, I'd just calculate the indices first and then use those to create the dense array. This allows one to simplify this to:
```
indices = tuple(arange(dim, dtype=dtype).reshape(
            shape[:i] + (dim,) + shape[i+1:]) for i, dim in enumerate(dimensions))
if sparse:
    return indices
else:
    return np.stack(np.broadcast_arrays(*indices))
```

Reply:
Does this work when `N == 0`?

Reply:
Unfortunately it doesn't, because `np.stack` needs at lease one array to stack and fails for `np.stack([])`, so it fails for the `dimensions==()` case. So this needs at least a special path for this case. I still like it more then the old solution, however it is a more severe code change overall.

Reply:
@Ablinne - multiple options:
1. `return np.stack(...) if N > 0 else np.array([], dtype=dtype)`
2. Stick with what you have.
3. Still make the empty array and loop with the insertions but after creating the tuple:
```
# TODO: once stack allows an empty tuple, this could be replaced with
# np.stack(np.broadcast_arrays(*indices))
res = empty((N,)+dimensions, dtype=dtype)
for r, index in zip(res, indices):
    r[...] = index
return res
```

Reply:
I am going to stick with the current version